beluga_system_tests/test/test_system.cpp

```diff
@@ -41,405 +31,225 @@
-        amcl_pf_from_map, "./bags/perfect_odometry", initial_pose, laser_info, "/map", "/odometry/ground_truth",
-        "/scan", "/odometry/ground_truth");
-  });
-  ret.emplace_back([]() {
-    InitialPose initial_pose{
-        Eigen::Vector3d{0.0, 2.0, 0.0},
-        Eigen::Matrix3d{
-            {0.125, 0.0, 0.0},
-            {0.0, 0.125, 0.0},
-            {0.0, 0.0, 0.04},
-        },
-    };
-    LaserScanInfo laser_info;
-    laser_info.laser_transform = Sophus::SE3d{Eigen::Quaterniond{1., 0., 0., 0.}, Eigen::Vector3d{0.28, 0., 0.}};
-    std::cout << laser_info.laser_transform.matrix() << std::endl;
-    laser_info.max_beam_count = 60;
-    laser_info.range_max = 100.;
-    laser_info.range_min = 0.;
-    return test_data_from_ros2bag<SE2dFromOdometryReader>(
-        likelihood_mcl_pf_from_map, "./bags/perfect_odometry", initial_pose, laser_info, "/map",
-        "/odometry/ground_truth", "/scan", "/odometry/ground_truth");
-  });
-  ret.emplace_back([]() {
-    InitialPose initial_pose{
-        Eigen::Vector3d{0.0, 2.0, 0.0},
-        Eigen::Matrix3d{
-            {0.125, 0.0, 0.0},
-            {0.0, 0.125, 0.0},
-            {0.0, 0.0, 0.04},
-        },
-    };
-    LaserScanInfo laser_info;
-    laser_info.laser_transform = Sophus::SE3d{Eigen::Quaterniond{1., 0., 0., 0.}, Eigen::Vector3d{0.28, 0., 0.}};
-    std::cout << laser_info.laser_transform.matrix() << std::endl;
-    laser_info.max_beam_count = 60;
-    laser_info.range_max = 100.;
-    laser_info.range_min = 0.;
-    return test_data_from_ros2bag<SE2dFromOdometryReader>(
-        beam_mcl_pf_from_map, "./bags/perfect_odometry", initial_pose, laser_info, "/map", "/odometry/ground_truth",
-        "/scan", "/odometry/ground_truth");
-  });
-  return ret;
+TEST_P(ParticleFilterTest, PerfectOdometryEstimatedPath) {
+  auto [map, datapoints, distribution] = get_perfect_odometry_data();
+  auto [particle_filter_params, motion_model, sensor_model_builder] = GetParam();
+  auto sensor_model = sensor_model_builder(map);
+  std::visit(
+      [map = std::move(map), distribution = std::move(distribution), datapoints = std::move(datapoints)]  //
+      (auto particle_filter_params, auto motion_model, auto sensor_model) mutable {
+        particle_filter_test(
+            particle_filter_params,   //
+            std::move(map),           //
+            std::move(motion_model),  //
+            std::move(sensor_model),  //
+            std::move(distribution),  //
+            std::move(datapoints));
+      },
+      particle_filter_params, std::move(motion_model), std::move(sensor_model));
 }
 
-INSTANTIATE_TEST_SUITE_P(BagFileTests, BelugaSystemTest, testing::ValuesIn(get_test_parameters()));
+INSTANTIATE_TEST_SUITE_P(
+    BagFileTest,
+    ParticleFilterTest,
+    testing::Combine(
+        testing::ValuesIn(get_particle_filter_params()),  //
+        testing::ValuesIn(get_motion_models()),           //
+        testing::ValuesIn(get_sensor_model_builders())));
```

Comment:
The diff obfuscates it a bit but the test is now amazingly straightforward. It all comes together nicely.